sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetColumn.java

```diff
@@ -0,0 +1,316 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.parquet;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+import com.google.common.base.Preconditions;
+import org.apache.spark.memory.MemoryMode;
+import org.apache.spark.sql.execution.vectorized.OffHeapColumnVector;
+import org.apache.spark.sql.execution.vectorized.OnHeapColumnVector;
+import org.apache.spark.sql.execution.vectorized.WritableColumnVector;
+import org.apache.spark.sql.types.ArrayType;
+import org.apache.spark.sql.types.DataType;
+import org.apache.spark.sql.types.DataTypes;
+import org.apache.spark.sql.types.MapType;
+import org.apache.spark.sql.types.StructType;
+
+/**
+ * Contains necessary information representing a Parquet column, either of primitive or nested type.
+ */
+final class ParquetColumn {
+  private final ParquetType type;
+  private final List<ParquetColumn> children;
+  private final WritableColumnVector vector;
+
+  /**
+   * repetition & definition levels
+   * these are allocated only for leaf columns; for non-leaf columns, they simply maintain
+   * references to that of the former.
+   */
+  private WritableColumnVector repetitionLevels;
+  private WritableColumnVector definitionLevels;
+
+  /** whether this column is primitive (i.e., leaf column) */
+  private final boolean isPrimitive;
+
+  /** reader for this column - only set if 'isPrimitive' is true */
+  private VectorizedColumnReader columnReader;
+
+  ParquetColumn(
+      ParquetType type,
+      WritableColumnVector vector,
+      int capacity,
+      MemoryMode memoryMode,
+      Set<ParquetType> missingColumns) {
+    DataType sparkType = type.sparkType();
+    if (!sparkType.sameType(vector.dataType())) {
+      throw new IllegalArgumentException("Spark type: " + type.sparkType() +
+        " doesn't match the type: " + vector.dataType() + " in column vector");
+    }
+
+    this.type = type;
+    this.vector = vector;
+    this.children = new ArrayList<>();
+    this.isPrimitive = type.isPrimitive();
+
+    if (missingColumns.contains(type)) {
+      vector.setAllNull();
+      return;
+    }
+
+    if (isPrimitive) {
+      repetitionLevels = allocateLevelsVector(capacity, memoryMode);
```

Comment:
Does the code handle situations when there are no repetition or definition levels for the type? Would it allocate vectors even when the column is null?

Reply:
If the column is null (i.e., a missing column), the code will exit at line 74 above, so it won't allocate vectors for these. I'm not quite sure what do you mean by "when there are no repetition or definition levels for the type", can you elaborate?

Reply:
I might have misunderstood the code, my question was about what happens when you read a primitive type that is non-null and non-repeated. Would the code still allocate the level vectors even if they are not necessarily needed? Again, it probably does not matter that much but maybe it is something we can clarify with a comment.

Reply:
@sadikovi ah I see what you mean. Good point. Let me see if I can remove these in those cases.

Reply:
Actually it's a bit tricky to remove these, although possible. Let me mark a TODO here and we can tackle it separately.

Reply:
Yes, I also thought it might be tricky to fix. It is fine to keep as is, I guess, especially if performance is good anyway.